### ISSUE TITLE
Console log shows the wrong destination file 

### ISSUE THREAD

**Issue by QJerryxjh** [facebook/react#22257]
in my browser console, log info go wrong, it should show real console.log file and correct line info, but now it always show "react_devtools_backend.js: 4049", link to source code show warning:
 "Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-unsafe-component-lifecycles for details.
* Move data fetching code or side effects to componentDidUpdate.
* If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
* Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.
Please update the following components: %s"
when i disable react developer tools, it goes correctly. but it goes wrong in 4.18.0 react developer tools. and my colleagues' chrome with react developer tools 4.17.0 is right

### Note from maintainers: https://github.com/facebook/react/issues/22257#issuecomment-916174962 is a workaround until this is fixed in React DevTools

**Comment by gaearon** [comment]
Can you try clicking the arrow to the left of the message? It should expand the full stack which gives you precise location of every frame including the one you’re looking for.

Also please show the screenshot of your DevTools settings.

**Comment by QJerryxjh** [comment]
@

**Comment by QJerryxjh** [comment]
@gaearon sry, it looks like i have a wrong devtools setting,thank you for you help

**Comment by mattywong** [comment]
Hey @gaearon, we seem to also be getting this issue on regular console logs (sourcemaps mapping to react_devtools_backend.js:4049). When the react devtools extension is disabled, sourcemaps show correctly. Webpack devtool is set to 'eval-source-map'. This seems to have happened quite recently as I don't recall seeing this before. Let me know if you need me to provide any other info.

Left: incognito with react devtools disabled, Right: with react devtools enabled

![image](https://user-images.githubusercontent.com/1550219/132461857-7672bb7a-d156-4ad1-940a-667896963f14.png)

**Comment by gaearon** [comment]
Not sure if anything changed here but reopening just in case.

**Comment by keul** [comment]
I can confirm: it seems that every `console.something` is captured by the extension.

![Schermata 2021-09-08 alle 10 53 14](https://user-images.githubusercontent.com/284924/132479185-7b2e6268-381b-48ca-bc92-60d7a3436edc.png)

**Comment by bvaughn** [comment]
Chiming in here to confirm that this reported behavior is intentional (although I understand the confusion or annoyance).

As of the [version 18 release](https://github.com/facebook/react/blob/main/packages/react-devtools/CHANGELOG.md#4180-september-1-2021), DevTools _always overrides the native console_ to either dim or suppress `StrictMode` double logging. (Before it only did it if you enabled a feature like component stacks.)

The major unfortunate drawback of this overriding is that it changes the location shown by the browser's console as reported in this issue. I requested better tooling support from browsers a couple of years ago because of this, but there hasn't been any action unfortunately:
https://github.com/whatwg/console/issues/163

There is a [related tc39 proposal](https://github.com/tc39/proposal-function-implementation-hiding) but I don't think it has made any progress since then.

**Comment by mattywong** [comment]
Perhaps a setting could be added to devtools to opt in to this new behaviour (or opt out)? Atleast till theres a way to fix both? Tbh i had no idea of the double logging issue or maybe had just never noticed it, and I would assume others would be in the same boat. The source map line change however is definitely noticeable and arguably a much more annoying issue (imo)

**Comment by gaearon** [comment]
Yes, we understand why the issue is annoying and will look into some way to resolve it.

**Comment by jzombie** [comment]
This is awesome.  I browsed the issue list, looking just for this, saw it as the first issue, subscribed to it while ago, and now saw the latest reply from @gaearon.

**Comment by allengordon011** [comment]
I've found a workaround using `console.info` instead of `console.log`. I hope that's helpful

**Comment by jzombie** [comment]
For what it's worth, I've found that doing things like this, allows you to override the console with a prefixer while retaining the original stack trace.  Not sure if that could be of any relevant help here, but I'm using it in a project for a custom logger, because I found that most logging overrides screw the stack trace.

console.log = Function.prototype.bind.call(
  console.log,
  console,
  "some-logging-prefix"
);

**Comment by bvaughn** [comment]
@jzombie That's a nice tip, but I don't think that offers a way for us to do anything more than inject a static parameter, does it? (I don't think it offers a way to run code, like we're doing in this case.)

**Comment by jzombie** [comment]
@bvaughn I've only used it to inject a static parameter, but perhaps it could run some code with a bit of fiddling.

**Comment by jzombie** [comment]
i.e. maybe if you used some sort of object property getter to render the string, and ran the code inside the getter... ?

**Comment by bvaughn** [comment]
If you have any example code/idea you could share, we'd be happy to take a look. 😁 

I think a getter would be evaluated at the time you define the `call` binding, rather than at the time you called the console method...but maybe you have an idea? :)

**Comment by jzombie** [comment]
I tried a couple of different approaches because I was really hoping it would work, and couldn't get beyond the initial value.  

I'd love to pursue this, this afternoon, but my job would probably not be happy about it, because it's completely unrelated to what I'm supposed to be working on.

**Comment by bvaughn** [comment]
As a temporary workaround for folks affected by this, you might try adding the `react_devtools_backend.js` script to the browser's "ignore list":
![image](https://user-images.githubusercontent.com/29597/132709425-0ae04aa3-3176-437a-8d99-bfb0b30537c9.png)

**Comment by vzaidman** [comment]
In Chrome, a workaround would be adding the script that overrides the console to the "ignore list" (right-click on the script in dev-tools and add to ignore list).

It has some drawbacks, because "ignore list" does a little more than just that, but it works nicely for me for now.

**Comment by JuraJuki** [comment]
Any updates on this issue?

**Comment by jzombie** [comment]
@bvaughn   I experimented a little and this is the closest I could get to make something work, where each call to console.log would call alert() with a new index, while retaining the original stack trace.

Maybe it could be useful, or maybe not.

![console override](https://user-images.githubusercontent.com/1139657/133622639-89990c1b-598e-48c6-a757-5e678fe195ba.png)


`<script>
  // Original console log
  const oLog = window.console.log

  const console = {
    // Inherit window console
    ...window.console,

    idx: -1,

    // Overridden console log
    get log() {
      // Alert idx on each log call
	    ++this.idx;
      alert(this.idx)   

      // Return original console log with original stack trace
      return oLog
    }
  }

	console.log("hello"); // alert(0); console.runner.html:22
	console.log("hello"); // alert(1); console.runner.html:23
	console.log("hello"); // alert(2); console.runner.html:24
</script>
`

**Comment by jzombie** [comment]
One more example showing the console log itself changing per call.

![console runner 2](https://user-images.githubusercontent.com/1139657/133625887-6f78329f-5b96-4046-986c-b0cdd3406335.png)

```
<script>
  const console = {
    idx: -1,

    // Overridden console log
    get log() {
      ++this.idx;

      // Return original console log with original stack trace
      return window.console.log.bind(window.console, `Iteration: ${this.idx}`);
    },
  };

  console.log("hello"); // "Iteration: 0 hello"; console.runner.html:14
  console.log("hello"); // "Iteration: 1 hello"; console.runner.html:15
  console.log("hello"); // "Iteration: 2 hello"; console.runner.html:16
</script>
```

**Comment by gaearon** [comment]
@JuraJuki If there were updates on this issue, they would be… on this issue. Please refrain from comments like this — they create notification noise for other people and don’t bring the conversation forward.

There’s ongoing work on this. When it’s done we’ll close the issue.

**Comment by EricRabil** [comment]
Another option to work around this without disabling the backend is to `Object.freeze(console)` - this is not a long-term fix but does let you keep the other parts of the backend.

**Comment by guncha** [comment]
Sean posted a workaround for the issue on SO for those that do not care for this "feature".

https://stackoverflow.com/a/69153985/357978

> I was able to fix it in my console by adding the file to my Framework Ignore List in Chrome. Steps:
> `Console -> Gear icon (top right) -> Settings -> Framework Ignone List.`
> Add react_devtools_backend.js to the list and the console.log should start showing the correct mapping again.

![image](https://user-images.githubusercontent.com/679632/135079408-dbd542d6-4490-429c-8a6b-cdbd819d74ab.png)

**Comment by jzombie** [comment]
I find it really ironic that the very same people who created this amazing library, and then also make the underlying functionality which drive hooks, can't wrap the browser console and not lose the stack trace.

Perhaps, just don't wrap the console at all, or find a library that does it and use it instead?

**Comment by bvaughn** [comment]
The solution described above:
https://github.com/facebook/react/issues/22257#issuecomment-915250655

Was released yesterday with DevTools version 4.19.0:
https://github.com/facebook/react/blob/main/packages/react-devtools/CHANGELOG.md#4190-september-29-2021

**Comment by guncha** [comment]
If you agree that this is an issue when it happens all of the time, why would you change it so that it only happens some of the time?

**Comment by bvaughn** [comment]
@guncha Your comment is missing some important context:
1. There is no way to tap into the console API and add information without triggering this observable effect. It's unfortunate but no browser API exists for this.
2. `StrictMode` double console logging has long been controversial (see #21783). Overriding the console to dim or hide the second logged message (#22030) is the best compromise we've been able to come up with.

**Comment by petetnt** [comment]
Any ETA for 4.19. (or 4.19.1.) landing to the Chrome web store?

**Comment by bvaughn** [comment]
No. I have no insight into Chrome's review process.

**Comment by petetnt** [comment]
@bvaughn that's understandable, no worries. Thanks for your hard work, React Dev Tools is an invaluable tool ❤️

**Comment by bvaughn** [comment]
4.19.0 is finally through review which unblocks 4.19.1 from being _submitted_. (Not sure how long it will take them to approve it.) I'm on PTO today b'c it's my birthday but I've asked if someone else on the team would submit it.

**Comment by yairEO** [comment]
still happens in version `4.22.0`

**Comment by gaearon** [comment]
@yairEO

What exactly "still happens"? This is not a helpful report because we don't know which part you're referring to.

I have verified that the extension does not clobber the log location for the first render. It does clobber the location for the second render in React 18, but that's expected. Hopefully the first location should always be enough.

<img width="513" alt="Screenshot 2022-03-30 at 18 34 36" src="https://user-images.githubusercontent.com/810438/160896835-0badd913-7036-4457-93e6-3a3557bad821.png">

I am going to lock this issue to avoid confusion. The original problem _has_ been resolved. If you see a problem, please file a new issue.
